### ISSUE TITLE
BUG: A server randomly disconnects and prevents a player from logging in

### ISSUE THREAD

**Issue by leolion3** [DarkflameUniverse/DarkflameServer#833]
### Make sure you've done the following:

- [X] I have checked that this issue has not already been reported.

- [X] I have validated that this issue is not a syntax error of either MySQL or SQLite.

- [X] I have pulled the latest version of the main branch of DarkflameServer and have confirmed that the issue exists there.

### DarkflameServer Version

848c0669248c506e5b916f52b9fec436ac655548

### Issue Description

Wo I've been operating my server for a couple months now, and it runs smoothly most of the time.

I've only noticed that when ran for multiple days the AuthServer seems to randomly disconnect at some point (without any error trace whatsoever) and not reconnect.

I have tried building around the issue by not allowing the AuthServer to shut down when it gets disconnected, in the hopes of it reconnecting, as I thought it might be my DNS that is sometimes inaccessible (.tk domains suck). This can be seen im my fork: https://github.com/leolion3/DarkflameServer

I've spent numerous hours debugging and refactoring code to no avail.

### Reproduction Steps

This is the tricky bit. Just run the server for a couple days and see if it's still up a week later.

### Expected Behavior

The AuthServer should not disconnrct in the first place or should reconnect if a disconnect was detected.

### Environment

Ubuntu 20.04

Hardware specs (physical machine at home):
- i7-7700HQ
- 16 GB RAM
- 512GB NVME SSD
- GTX 970m

Network infrastructure: Unifi OS, 100MBit connection (down/up)

**Comment by MustacheMan2** [comment]
Nice, this is the same issue I've had since launch but I haven't ever bothered to write a report. Originally I thought Account Manager/Nexus Dash was interfering because my server only has 2 GB ram and likely not a very strong CPU but as seen here that's clearly not the case. Pretty much the only thing that demands a server restart every couple weeks or so, otherwise uptime could be always.

**Comment by EmosewaMC** [comment]
Nice to see our server is not the only one with the issue.  We have literally no idea whats going on.  No one has been able to pin down why auth just dies out of the blue, or whether its even an auth problem at all.  Master should be able to detect when auth goes down in any case and spawn a new auth server in the worst case

**Comment by EmosewaMC** [comment]
I have changed the title to be more indicative of the issue.  It has not been imperically proven that Auth is the issue, people just cant login through auth and assume its an auth issue.  Unless you have 100% irrefutable proof, its not an auth issue

**Comment by MustacheMan2** [comment]
> I have changed the title to be more indicative of the issue. It has not been imperically proven that Auth is the issue, people just cant login through auth and assume its an auth issue. Unless you have 100% irrefutable proof, its not an auth issue

Well there's not a ton of information on it but I'm seeing in my logs that auth just loses connection with master (but this isn't mentioned by master, however its logs end not long after auth shuts down).

**Comment by EmosewaMC** [comment]
> > I have changed the title to be more indicative of the issue. It has not been imperically proven that Auth is the issue, people just cant login through auth and assume its an auth issue. Unless you have 100% irrefutable proof, its not an auth issue
> 
> Well there's not a ton of information on it but I'm seeing in my logs that auth just loses connection with master (but this isn't mentioned by master, however its logs end not long after auth shuts down).

Can you post your logs of the incident please? Auth Chat Master and zone 0 logs would be great to aid in the triage process!

**Comment by MustacheMan2** [comment]
[AuthServer_1667920496.log](https://github.com/DarkflameUniverse/DarkflameServer/files/10046184/AuthServer_1667920496.log)
[MasterServer_1667920495.log](https://github.com/DarkflameUniverse/DarkflameServer/files/10046185/MasterServer_1667920495.log)
[WorldServer_0_0_1667920496.log](https://github.com/DarkflameUniverse/DarkflameServer/files/10046186/WorldServer_0_0_1667920496.log)

Alright I guess I was a little wrong with the ordering of these particular logs, looks like Auth actually shut down _after_ WorldServer 0 and Master stopped logging, and no crash logs are present at all. Pretty strange.

**Comment by EmosewaMC** [comment]
How do you run your server?  What commands specifically do you use to stop and start the server?
Can you tell me what signals you send to your server when it is shutting down?  Ots looking like master is killed on your end via an unsupported signal which causes no server to shutdown properly.

**Comment by leolion3** [comment]
I run my server using a custom run script found [here](gist.github.com/leolion3/ccf654ab60c8e110c65ef948da6af461). Note that I run my servers without sudo for security reasons (port 1001 usage is allowed before the servers are executed.

The shutdown procedure is always the same: sometime after no one has been online (mostly around 20 minutes to one hour) the auth server suddenly loses connection to the master server out of the blue and then the mysql connection is destroyed and the auth server shuts down. The worldservers are already offline at that point and so are the chatservers. Only the masterserver continues running.

> How do you run your server? What commands specifically do you use to stop and start the server? Can you tell me what signals you send to your server when it is shutting down? Ots looking like master is killed on your end via an unsupported signal which causes no server to shutdown properly.

**Comment by EmosewaMC** [comment]
all other world servers, except world zero.  Please check your world zero logs and confirm that world zero also does not loose connection at the same time as auth

**Comment by leolion3** [comment]
> all other world servers, except world zero. Please check your world zero logs and confirm that world zero also does not lose connection at the same time as auth

Just checked my server and it happened again. Heres the full log:

```
[13-11-22 21:47:46] [WorldServer] Deleting player 1152921508901814274
[13-11-22 21:47:46] [Player] Deleted player
[13-11-22 21:47:46] [TotalTimePlayed] Time since last save: 0
[13-11-22 21:47:46] [Character] Saved character to Database in: 0.006009s
[13-11-22 21:47:46] [UserManager] Deleted user 1
[13-11-22 21:47:46] [PlayerContainer] Removed user: 1152921508901814274
[13-11-22 22:29:02] [WorldServer] Saving map 1000 instance 1
[13-11-22 22:29:02] [WorldServer] ALL DATA HAS BEEN SAVED FOR ZONE 1000 INSTANCE 1!
[13-11-22 22:29:02] [WorldServer] Shutdown complete, zone (1000), instance (1)
[13-11-22 22:29:02] [Database] Destroying MySQL connection from WorldServer!
[13-11-22 22:29:02] [MasterServer] Got shutdown response from zone 1000 clone 0 instance 1 port 3003
[13-11-22 22:29:02] [MasterServer] A server has disconnected
[13-11-22 22:29:02] [MasterServer] Actually disconnected from zone 1000 clone 0 instance 1 port 3003
[13-11-22 22:29:14] [ChatServer] A server has disconnected, erasing their connected players from the list.
[13-11-22 22:31:46] [WorldServer] Saving map 1200 instance 2
[13-11-22 22:31:46] [WorldServer] ALL DATA HAS BEEN SAVED FOR ZONE 1200 INSTANCE 2!
[13-11-22 22:31:46] [WorldServer] Shutdown complete, zone (1200), instance (2)
[13-11-22 22:31:46] [Database] Destroying MySQL connection from WorldServer!
[13-11-22 22:31:46] [MasterServer] Got shutdown response from zone 1200 clone 0 instance 2 port 3006
[13-11-22 22:31:46] [MasterServer] A server has disconnected
[13-11-22 22:31:46] [MasterServer] Actually disconnected from zone 1200 clone 0 instance 2 port 3006
[13-11-22 22:32:00] [ChatServer] A server has disconnected, erasing their connected players from the list.
[15-11-22 23:37:00] [dServer] Lost our connection to master, shutting DOWN!
[15-11-22 23:37:01] [dServer] Lost our connection to master, shutting DOWN!
[15-11-22 23:37:03] [dServer] Lost our connection to master, shutting DOWN!
[15-11-22 23:37:03] [WorldServer] Game loop running but no connection to master for 10 frames, shutting down
[15-11-22 23:37:03] [WorldServer] Saving map 0 instance 0
[15-11-22 23:37:03] [WorldServer] ALL DATA HAS BEEN SAVED FOR ZONE 0 INSTANCE 0!
[15-11-22 23:37:03] [WorldServer] Shutdown complete, zone (0), instance (0)
[15-11-22 23:37:03] [Database] Destroying MySQL connection from WorldServer!
[15-11-22 23:37:07] [Database] Destroying MySQL connection from ChatServer!
[20-11-22 00:44:48] [AuthPackets] Received client version: 171022
[20-11-22 00:44:53] [AuthPackets] Set sessionKey: 846930886 for user LeoLion
[20-11-22 00:45:08] [AuthPackets] Received client version: 171022
[20-11-22 00:45:13] [AuthPackets] Set sessionKey: 1681692777 for user LeoLion
```

The client gets stuck either on `Handshaking with Authentication` or on `Submitting username`. 

![image](https://user-images.githubusercontent.com/45939246/202876051-0cca694c-9b83-478c-93cd-d366a2f39f47.png)

As you see, the server was running fine from the 13th till the 15th, which is when the servers crashed. There was absolutely no interaction with the game server between these dates. Nobody logged in or out.

The server is running in a screen instance, which is still live.

**Comment by EmosewaMC** [comment]
What is interesting is that world zero actually shutdown.  This only ever happens if world zero loses connection with master.  This is more likely an issue with master and not auth, as we have thought.  What is odd is that auth _doesnt_ shutdown actually, but keeps running.

**Comment by EmosewaMC** [comment]
Had this happen again.  All servers died due to disconnection from master, however Master was still running.  Very strange behavior since even if master is in an infinite loop, the server will not lose connection to other servers.  We should figure out if this is a disconnection or a lost connection (very likely the latter since the former is a message sent on RakPeer shutdown.

**Comment by leolion3** [comment]
> Had this happen again. All servers died due to disconnection from master, however Master was still running. Very strange behavior since even if master is in an infinite loop, the server will not lose connection to other servers. We should figure out if this is a disconnection or a lost connection (very likely the latter since the former is a message sent on RakPeer shutdown.

Any idea where the bug might lie? I've been refactoring some code but can't seem to find the origin of the error.

**Comment by EmosewaMC** [comment]
The very last thing the master server did just before all other servers lost connection was an instance transfer callback.  Whether or not the bug is actually there is another thing.  I've personally got no idea what's going on.

**Comment by leolion3** [comment]
> The very last thing the master server did just before all other servers lost connection was an instance transfer callback. Whether or not the bug is actually there is another thing. I've personally got no idea what's going on.

Alright I'll keep looking if I can find out what's happening. Could you increase the priority of the issue? It seems like a critical one considering it affects the availability of hosted servers.

**Comment by EmosewaMC** [comment]
Nope.  It only affects long running servers and is fixed with a restart.  This will remain low due to its infrequency and how easy the fix is.

**Comment by LunarCyber** [comment]
any updates on how to fix this? just curious because I feel like I wasted time for setting my own up to encounter it losing connection to master..

**Comment by EmosewaMC** [comment]
Just restart your server to fix the bug.

**Comment by LunarCyber** [comment]
> Just restart your server to fix the bug.

never really dealt with servers only like stuff for Kali Linux for scambaiting so I would if I knew how to sadly

**Comment by leolion3** [comment]
> > Just restart your server to fix the bug.
> 
> never really dealt with servers only like stuff for Kali Linux for scambaiting so I would if I knew how to sadly

Just set up a cron job that runs as sudo and executes the command `pkill WorldServer && pkill ChatServer && pkill AuthServer && pkill MasterServer && /path/to/build/MasterServer` or something similar. You can edit the crontab (cron job schedule) using `sudo crontab -e` (the above commands should all be executed as sudo as otherwise you'll get permission denied errors when using pkill as the server was previously ran as sudo)

**Comment by EmosewaMC** [comment]
I heavily advise against using any kill signals towards a server unless it is truly stuck.  Always use sig_int of sig_term.  If you use kill the server will not save.

**Comment by aronwk-aaron** [comment]
We've found the root issue of this problem and will work to address it:
`[SocketLayer] Returning error code -1 in SocketLayer due to bug with Zone Alarm.`

**Comment by EmosewaMC** [comment]
This is now addressed on main now with #885 .
